### PR TITLE
Upgraded to libxsmm 1.7

### PR DESCRIPTION
**Pull request by benoitsteiner**

### CODE REVIEW

**Comment by benoitsteiner** [comment]
@hfp Can you sign the CLA so that we can merge this in TensorFlow ?

**Comment by gunan** [comment]
Jenkins, test this please.

**Comment by bhack** [comment]
Is there any doc location where we documented on which cases it is reccomended to enable LIBXSMM?

**Comment by benoitsteiner** [comment]
@bhack We're planning to enable libxsmm by default whenever it makes sense. The codebase just needs to mature a little bit more before we do so.

**Comment by hfp** [comment]
> Can you sign the CLA so that we can merge this in TensorFlow ?

I signed it

**Comment by vrv** [comment]
You have to re-do the commits with the correct email - your git commit shows "From: Hans Pabst <hfp@users.noreply.github.com>"

**Comment by vrv** [comment]
Also, please either rebase or squash, there's some superfluous merge commits in here.  (Can probably squash and merge, whoever is on rotation).

**Comment by hfp** [comment]
I signed it
Intel Semiconductor AG
Registered No. 020.30.913.786-7
Registered Office: Badenerstrasse 549, 8048 Zurich, Switzerland

**Comment by rmlarsen** [comment]
I'm afraid we cannot accept the commit if the CLA signature doesn't match the email in the commit. Can you either recommit or sign with the correct email?

**Comment by vrv** [comment]
Benoit, maybe you can edit the git commit in your PR to use the email he registered.

**Comment by benoitsteiner** [comment]
I've edited the commit and replaced the hfp@users.noreply.github.com email address with hans.pabst@intel.com. The CLA bot is now happy.

File: tensorflow/core/kernels/conv_ops.cc [hunk 1]
```diff
@@ -204,12 +204,15 @@ class LaunchXsmmConvOp<CPUDevice, float> {
...
-    desc.filter_format = LIBXSMM_DNN_CONV_FORMAT_LIBXSMM;
+    desc.buffer_format = LIBXSMM_DNN_TENSOR_FORMAT_NHWC;
+    desc.filter_format = LIBXSMM_DNN_TENSOR_FORMAT_LIBXSMM;
     desc.fuse_ops = LIBXSMM_DNN_CONV_FUSE_NONE;
     desc.options = LIBXSMM_DNN_CONV_OPTION_NONE;
-    desc.datatype_in = LIBXSMM_DNN_DATATYPE_F32;
-    desc.datatype_out = LIBXSMM_DNN_DATATYPE_F32;
+    desc.datatype = LIBXSMM_DNN_DATATYPE_F32;
+
+    if (!CanUseXsmmConv2D(desc, data_format)) {
```
Comment (gunan) [inline tensorflow/core/kernels/conv_ops.cc]: The same if block duplicated?

File: tensorflow/core/kernels/sparse_matmul_op.cc [hunk 2]
```diff
@@ -896,6 +897,8 @@ class LibxsmmSparseMatMul {
       } else {
         std::unique_ptr<TensorInfoCacheEntry> e{
             new TensorInfoCacheEntry{M, K, N, max_threads, {}, nullptr}};
+        // setup scoped allocator, which uses cpu_allocator() for this scope
+        const libxsmm_tf_allocator<libxsmm_scratch_allocator> tf_allocator;
```
Comment (gunan) [inline tensorflow/core/kernels/sparse_matmul_op.cc]: I'm sure I'm missing something, but possibly unused variable?

File: tensorflow/core/kernels/xsmm_conv2d.cc [hunk 3]
```diff
@@ -158,11 +159,11 @@ struct HashFunction{
  
 class handles{
   public:
-    libxsmm_dnn_conv_handle* find( const libxsmm_dnn_conv_desc_wrap &w) {
-      std::unordered_map<libxsmm_dnn_conv_desc_wrap , libxsmm_dnn_conv_handle*, HashFunction>::iterator i = libxsmm_handles.find(w);
+    libxsmm_dnn_layer* find( const libxsmm_dnn_conv_desc_wrap &w) {
+      std::unordered_map<libxsmm_dnn_conv_desc_wrap , libxsmm_dnn_layer*, HashFunction>::iterator i = libxsmm_handles.find(w);
```
Comment (gunan) [inline tensorflow/core/kernels/xsmm_conv2d.cc]: line longer than 80 chars?

File: tensorflow/core/kernels/xsmm_conv2d.cc [hunk 4]
```diff
@@ -224,10 +226,8 @@ static bool CallLibxsmmConvGeneric(OpKernelContext* ctx,
 
   int blocksifm = desc.C%ifmblock ==0 ? desc.C/ifmblock :desc.C/ifmblock + 1;           
   int blocksofm = desc.K%ofmblock ==0 ? desc.K/ofmblock :desc.K/ofmblock + 1;
-  float *native_filter = (float*)libxsmm_aligned_malloc( blocksofm*blocksifm*desc.R*desc.S*ifmblock*ofmblock*sizeof(float), 2097152);
- 
+  float *native_filter = (float*)libxsmm_aligned_scratch( blocksofm*blocksifm*desc.R*desc.S*ifmblock*ofmblock*sizeof(float), 2097152);
```
Comment (gunan) [inline tensorflow/core/kernels/xsmm_conv2d.cc]: line length?

File: tensorflow/core/kernels/conv_ops.cc (hunk 1 above)
Comment (benoitsteiner) [inline tensorflow/core/kernels/conv_ops.cc]: Fixed. Thanks for catching this

File: tensorflow/core/kernels/sparse_matmul_op.cc (hunk 2 above)
Comment (benoitsteiner) [inline tensorflow/core/kernels/sparse_matmul_op.cc]: As far as I can tell, this ensures that all the calls that libxsmm makes to malloc use the tf_allocator at least for the duration of the tf_allocator variable.

File: tensorflow/core/kernels/xsmm_conv2d.cc (hunk 3 above)
Comment (benoitsteiner) [inline tensorflow/core/kernels/xsmm_conv2d.cc]: Fixed

File: tensorflow/core/kernels/xsmm_conv2d.cc (hunk 4 above)
Comment (benoitsteiner) [inline tensorflow/core/kernels/xsmm_conv2d.cc]: Fixed

**Review by gunan — APPROVED** [review]

**Pull request merged**